src/Cli/dotnet/commands/dotnet-test/TestingPlatformCommand.cs

```diff
@@ -0,0 +1,170 @@
+        private readonly ConcurrentDictionary<string, TestApplication> _testApplications = [];
+        private readonly PipeNameDescription _pipeNameDescription = NamedPipeServer.GetPipeName(Guid.NewGuid().ToString("N"));
+        private readonly CancellationTokenSource _cancellationToken = new();
+
+        private Task _namedPipeConnectionLoop;
+        private string[] _args;
+
+        public TestingPlatformCommand(string name, string description = null) : base(name, description)
+        {
+            TreatUnmatchedTokensAsErrors = false;
+        }
+
+        public int Run(ParseResult parseResult)
+        {
+            _args = parseResult.GetArguments().Except(parseResult.UnmatchedTokens).ToArray();
+
+            VSTestTrace.SafeWriteTrace(() => $"Wait for connection(s) on pipe = {_pipeNameDescription.Name}");
+            _namedPipeConnectionLoop = Task.Run(async () => await WaitConnectionAsync(_cancellationToken.Token));
+
+            bool containsNoBuild = parseResult.UnmatchedTokens.Any(token => token == CliConstants.NoBuildOptionKey);
+
+            if (containsNoBuild)
+            {
+                ForwardingAppImplementation mSBuildForwardingApp = new(GetMSBuildExePath(), ["-t:_GetTestsProject", $"-p:GetTestsProjectPipeName={_pipeNameDescription.Name}", "-verbosity:q"]);
+                int getTestsProjectResult = mSBuildForwardingApp.Execute();
+            }
+            else
+            {
+                string msBuildArgs = TryGetMSBuildArgs(parseResult, out string[] result) ?
+                    string.Join(" ", result) :
+                    string.Empty;
+
+                BuildCommand buildCommand = BuildCommand.FromArgs(["-t:_BuildTestsProject;_GetTestsProject", msBuildArgs, $"-p:GetTestsProjectPipeName={_pipeNameDescription.Name}", "-verbosity:q"]);
+                int buildResult = buildCommand.Execute();
+            }
+
+            // Above line will block till we have all connections and all GetTestsProject msbuild task complete.
+            Task.WaitAll([.. _taskModuleName]);
+            Task.WaitAll([.. _testsRun]);
+            _cancellationToken.Cancel();
+            _namedPipeConnectionLoop.Wait();
+
+            return 0;
+        }
+
+        private static bool TryGetMSBuildArgs(ParseResult parseResult, out string[] msbuildArgs)
+        {
+            var hasMSBuildArgs = parseResult.UnmatchedTokens.Any(token => token == CliConstants.MSBuildOptionKey);
+            msbuildArgs = hasMSBuildArgs ? parseResult.UnmatchedTokens.SkipWhile(token => token != CliConstants.MSBuildOptionKey).Skip(1).ToArray() : [];
+            return hasMSBuildArgs;
+        }
+
+        private async Task WaitConnectionAsync(CancellationToken token)
+        {
+            try
+            {
+                while (true)
+                {
+                    NamedPipeServer namedPipeServer = new(_pipeNameDescription, OnRequest, NamedPipeServerStream.MaxAllowedServerInstances, token);
```

Comment:
Since this is a new IPC, is there a threat model/review that's been done?

Reply:
Not at the moment

@drognanar for the json-rpc protocol have you done the threat model and review with someone?